common/packet/src/packet.cpp

```diff
@@ -0,0 +1,239 @@
+}
+
+Packet::~Packet() {
+    this->buffer = std::vector<u_int8_t>();
+}
+
+void Packet::write_byte(char data){
+    u_int8_t toAdd = u_int8_t(data);
+    this->buffer.push_back(toAdd);
+}
+
+// 32 bits
+// ref: https://stackoverflow.com/questions/6499183/converting-a-uint32-value-into-a-uint8-array4
+void Packet::write_int(int data){
+    union32.i = data;
+    u_int32_t bytes = union32.l;
+    bytes = htonl(bytes);
+
+    /* Equivalent code:
+    u_int8_t byte_array[4];
+    byte_array[0] = bytes >> 24;
+    byte_array[1] = bytes >> 16;
+    byte_array[2] = bytes >>  8;
+    byte_array[3] = bytes;
+
+
+    this->buffer.push_back(byte_array[0]);
+    this->buffer.push_back(byte_array[1]);
+    this->buffer.push_back(byte_array[2]);
+    this->buffer.push_back(byte_array[3]);*/
+
+    // store 4 bytes in the buffer
+    for(int i = 1; i <= 4; i++){
+        this->buffer.push_back((u_int8_t)(bytes >> 8*(4-i)));
+    }
+
+}
+
+// 32 bits
+void Packet::write_float(float data){
+    union32.f = data;
+    u_int32_t bytes = union32.l;
+    bytes = htonl(bytes);
+
+    // store 4 bytes in the buffer
+    for(int i = 1; i <= 4; i++){
+        this->buffer.push_back((u_int8_t)(bytes >> 8*(4-i)));
+    }
+
+}
+
+// ref: https://stackoverflow.com/questions/809902/64-bit-ntohl-in-c
+// ntoh function for 64 bit input
+u_int64_t
+ntoh64(const u_int64_t *input)
+{
+    u_int64_t rval;
+    u_int8_t *data = (u_int8_t *)&rval;
+
+    data[0] = *input >> 56;
+    data[1] = *input >> 48;
+    data[2] = *input >> 40;
+    data[3] = *input >> 32;
+    data[4] = *input >> 24;
+    data[5] = *input >> 16;
+    data[6] = *input >> 8;
+    data[7] = *input >> 0;
+
+    return rval;
+}
+
+// 64 bits
+void Packet::write_double(double data){
+    union64.f = data;
+    u_int64_t bytes = union64.l;
+
+    //std::cout << "double is: " << data << " u_int64 is: " << bytes;
+
+    bytes = ntoh64(&bytes);
+
+    //std::cout << " sending: " << bytes << std::endl;
+
+    // store 8 bytes in the buffer
+    for(int i = 1; i <= 8; i++){
+        u_int8_t toWrite = (u_int8_t)(bytes >> 8*(8-i));
+        this->buffer.push_back(toWrite);
+    }
+}
+
+// floats are 32 bits
+// store x, y, z
+void Packet::write_vec3(glm::vec3 data){
+    // write each element of the data vector
+    write_float(data.x);
+    write_float(data.y);
+    write_float(data.z);
+}
+
+int Packet::read_byte(char* dest){
+    if(this->buffer.size() == 0){
+        return -1;
+    }
+
+    u_int8_t readElem = this->buffer.front();
+    this->buffer.erase(this->buffer.begin());
+    *dest = char(readElem);
+    
+    return 1;
+}
+
+int Packet::read_int(int* dest){
+    int int8_per_int = 4;
+
+    if(this->buffer.size() < int8_per_int){
+        return -1;
+    }
+
+    // read 4 bytes from the buffer
+    u_int32_t readData[int8_per_int];
+    for(int i = 0; i < int8_per_int; i++){
+        readData[i] = this->buffer.front();
+        this->buffer.erase(this->buffer.begin());
```

Comment:
```suggestion
buffer.pop_front();
```
Don't need `this` as well